pkg/chunk/cassandra/storage_client.go

```diff
@@ -0,0 +1,219 @@
+package cassandra
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/gocql/gocql"
+
+	"github.com/weaveworks/cortex/pkg/chunk"
+)
+
+const (
+	maxRowReads = 100
+)
+
+// Config for a StorageClient
+type Config struct {
+	address           string
+	keyspace          string
+	replicationFactor int
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.address, "cassandra.address", "", "Address of Cassandra instances.")
+	f.StringVar(&cfg.keyspace, "cassandra.keyspace", "", "Keyspace to use in Cassandra.")
+	f.IntVar(&cfg.replicationFactor, "cassandra.replication-factor", 1, "Replication factor to use in Cassandra.")
+}
+
+func (cfg *Config) session() (*gocql.Session, error) {
+	if err := cfg.createKeyspace(); err != nil {
+		return nil, err
+	}
+
+	cluster := gocql.NewCluster(cfg.address)
+	cluster.Keyspace = cfg.keyspace
+	cluster.Consistency = gocql.Quorum
+	cluster.BatchObserver = observer{}
+	cluster.QueryObserver = observer{}
+
+	return cluster.CreateSession()
+}
+
+// createKeyspace will create the desired keyspace if it doesn't exist.
+func (cfg *Config) createKeyspace() error {
+	cluster := gocql.NewCluster(cfg.address)
+	cluster.Keyspace = "system"
+	cluster.Timeout = 20 * time.Second
+	session, err := cluster.CreateSession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	return session.Query(fmt.Sprintf(
+		`CREATE KEYSPACE IF NOT EXISTS %s
+		 WITH replication = {
+			 'class' : 'SimpleStrategy',
+			 'replication_factor' : %d
+		 }`,
+		cfg.keyspace, cfg.replicationFactor)).Exec()
+}
+
+// storageClient implements chunk.storageClient for GCP.
+type storageClient struct {
+	cfg       Config
+	schemaCfg chunk.SchemaConfig
+	session   *gocql.Session
+}
+
+// NewStorageClient returns a new StorageClient.
+func NewStorageClient(cfg Config, schemaCfg chunk.SchemaConfig) (chunk.StorageClient, error) {
+	session, err := cfg.session()
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageClient{
+		cfg:       cfg,
+		schemaCfg: schemaCfg,
+		session:   session,
+	}, nil
+}
+
+func (s *storageClient) Close() {
+	s.session.Close()
+}
+
+type writeBatch struct {
+	b *gocql.Batch
+}
+
+func (s *storageClient) NewWriteBatch() chunk.WriteBatch {
+	return writeBatch{
+		b: gocql.NewBatch(gocql.UnloggedBatch),
+	}
+}
+
+func (b writeBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	b.b.Query(fmt.Sprintf("INSERT INTO %s (hash, range, value) VALUES (?, ?, ?)", tableName),
```

Comment:
Nit/Performance: what values can `tableName` take? Asking as I'm wondering if this could be turned into a prepared statement to save a bit on the streaming of the query for every insert.

Reply:
I guess this actually happens transparently behind the scene: "_Automatic query preparation_" (apparently via a LRU cache).

Reply:
I originally had the table name as a ?, but turns out that not allowed.  The prepared statement cache will work nicely here though, its mostly only a few tables in use at once.

Reply:
> _The prepared statement cache will work nicely here though, its mostly only a few tables in use at once._

Excellent! :+1: